tensorflow_probability/python/experimental/mcmc/sample_fold.py

```diff
@@ -133,3 +139,174 @@ def sample_fold(
+    kernel: An instance of `tfp.mcmc.TransitionKernel` which implements one step
+      of the Markov chain.
+    num_burnin_steps: Integer number of chain steps to take before starting to
+      collect results.
+      Default value: 0 (i.e., no burn-in).
+    num_steps_between_results: Integer number of chain steps between collecting
+      a result. Only one out of every `num_steps_between_samples + 1` steps is
+      included in the returned results.  The number of returned chain states is
+      still equal to `num_results`.  Default value: 0 (i.e., no thinning).
+    trace_fn: A callable that takes in the current chain state and the previous
+      kernel results and return a `Tensor` or a nested collection of `Tensor`s
+      that is then traced along with the chain state.
+    return_final_kernel_results: If `True`, then the final kernel results are
+      returned alongside the chain state and the trace specified by the
+      `trace_fn`.
+    parallel_iterations: The number of iterations allowed to run in parallel. It
+      must be a positive integer. See `tf.while_loop` for more details.
+    seed: Optional, a seed for reproducible sampling.
+    name: Python `str` name prefixed to Ops created by this function.
+      Default value: `None` (i.e., 'experimental_mcmc_sample_chain').
+
+  Returns:
+    checkpointable_states_and_trace: if `return_final_kernel_results` is
+      `True`. The return value is an instance of
+      `CheckpointableStatesAndTrace`.
+    all_states: if `return_final_kernel_results` is `False` and `trace_fn` is
+      `None`. The return value is a `Tensor` or Python list of `Tensor`s
+      representing the state(s) of the Markov chain(s) at each result step. Has
+      same shape as input `current_state` but with a prepended
+      `num_results`-size dimension.
+    states_and_trace: if `return_final_kernel_results` is `False` and
+      `trace_fn` is not `None`. The return value is an instance of
+      `StatesAndTrace`.
+
+  #### References
+
+  [1]: Art B. Owen. Statistically efficient thinning of a Markov chain sampler.
+       _Technical Report_, 2017.
+       http://statweb.stanford.edu/~owen/reports/bestthinning.pdf
+  """
+  with tf.name_scope(name or 'experimental_mcmc_sample_chain'):
+    if not kernel.is_calibrated:
+      warnings.warn('supplied `TransitionKernel` is not calibrated. Markov '
+                    'chain may not converge to intended target distribution.')
+
+    if trace_fn is None:
+      trace_fn = lambda *args: ()
+      no_trace = True
+    else:
+      no_trace = False
+
+    if trace_fn is sample_chain.__defaults__[4]:
+      warnings.warn('Tracing all kernel results by default is deprecated. Set '
+                    'the `trace_fn` argument to None (the future default '
+                    'value) or an explicit callback that traces the values '
+                    'you are interested in.')
+
+    tracing_reducer = mcmc.TracingReducer(
```

Comment:
Could you import the module for TracingReducer directly above, instead of trying to go through the package?  IIRC, importing the package caused problems in Python 2.

Reply:
yes, absolutely! I made this PR before I asked about this weird spec so I didn't know about the py2 incompatibility